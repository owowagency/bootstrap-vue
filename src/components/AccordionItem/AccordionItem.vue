<template>
    <div class="accordion-item">
        <Component
            :is="headerTag"
            class="accordion-header"
            :class="headerClass"
        >
            <button
                class="accordion-button collapsed"
                :class="buttonClass"
                data-bs-toggle="collapse"
                :data-bs-target="`#${id}`"
            >
                <!-- @slot Displays the header of the accordion -->
                <slot name="header">
                    {{ header }}
                </slot>
            </button>
        </Component>

        <div
            :id="`${id}`"
            class="accordion-collapse collapse"
            :class="collapseClass"
            :data-bs-parent="parentId ? `#${parentId}` : undefined"
        >
            <div
                class="accordion-body"
                :class="bodyClass"
            >
                <!-- @slot Displays the body of the accordion -->
                <slot />
            </div>
        </div>
    </div>
</template>

<script lang="ts">
import {idProps} from '@/composables/useId';
</script>

<script lang="ts" setup>
defineProps({
    /**
     * The classes that will be forwarded to the body
     */
    bodyClass: {
        type: [String, Array, Object],
        default: undefined,
    },
    /**
     * The classes that will be forwarded to the button
     */
    buttonClass: {
        type: [String, Array, Object],
        default: undefined,
    },
    /**
     * The classes that will be forwarded to the collapse
     */
    collapseClass: {
        type: [String, Array, Object],
        default: undefined,
    },
    /**
     * The text that will be shown in the header
     */
    header: {
        type: String,
        default: undefined,
    },
    /**
     * The classes that will be forwarded to the header
     */
    headerClass: {
        type: [String, Array, Object],
        default: undefined,
    },
    /**
     * The html tag that will be used for the header
     */
    headerTag: {
        type: String,
        default: 'h2',
    },
    ...idProps,
    /**
     * The id of the parent `Accordion` component
     */
    parentId: {
        type: String,
        default: undefined,
    },
});
</script>

<docs>
```vue
<AccordionItem header="I'm the header">
    <div>I'm the body</div>
</AccordionItem>
```

## Imported Props

| Prop name | Description                   | Type   | Values | Default |
| --------- | ----------------------------- | ------ | ------ | ------- |
| id        | The id used for the component | string | -      | `uuid`  |
</docs>
