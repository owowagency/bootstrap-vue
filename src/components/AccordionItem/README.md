# AccordionItem

```vue
<AccordionItem header="I'm the header">
    <div>I'm the body</div>
</AccordionItem>
```

## Imported Props

| Prop name | Description                   | Type   | Values | Default |
| --------- | ----------------------------- | ------ | ------ | ------- |
| id        | The id used for the component | string | -      | `uuid`  |

## Props

| Prop name     | Description                                        | Type                  | Values | Default   |
| ------------- | -------------------------------------------------- | --------------------- | ------ | --------- |
| bodyClass     | The classes that will be forwarded to the body     | string\|array\|object | -      | undefined |
| buttonClass   | The classes that will be forwarded to the button   | string\|array\|object | -      | undefined |
| collapseClass | The classes that will be forwarded to the collapse | string\|array\|object | -      | undefined |
| header        | The text that will be shown in the header          | string                | -      | undefined |
| headerClass   | The classes that will be forwarded to the header   | string\|array\|object | -      | undefined |
| headerTag     | The html tag that will be used for the header      | string                | -      | 'h2'      |
| parentId      | The id of the parent `Accordion` component         | string                | -      | undefined |

## Slots

| Name    | Description                          | Bindings |
| ------- | ------------------------------------ | -------- |
| header  | Displays the header of the accordion |          |
| default | Displays the body of the accordion   |          |
