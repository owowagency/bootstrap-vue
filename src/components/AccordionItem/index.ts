import AccordionItem from './AccordionItem.vue';

export default AccordionItem;
