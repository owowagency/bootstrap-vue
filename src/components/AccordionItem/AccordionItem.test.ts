import AccordionItem from '.';
import {shallowMount} from '@vue/test-utils';

describe('template', () => {
    componentRenderTest(AccordionItem, {
        // Force an id to prevent snapshots from failing because of uuid.
        props: {id: 'dont-change-please'},
    });

    componentSlotRenderTest(AccordionItem);

    componentSlotRenderTest(AccordionItem, 'header');

    it('renders header', () => {
        const wrapper = shallowMount(AccordionItem, {
            props: {
                id: 'owow-accordion-item',
                header: 'find-me',
                headerTag: 'h5',
            },
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('renders given custom classes', () => {
        const wrapper = shallowMount(AccordionItem, {
            props: {
                id: 'owow-accordion-item',
                bodyClass: 'body-class',
                buttonClass: 'button-class',
                collapseClass: 'collapse-class',
                class: 'custom-class',
                headerClass: 'header-class',
            },
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('adds parent id to collapse', () => {
        const wrapper = shallowMount(AccordionItem, {
            props: {parentId: 'i-am-your-father'},
        });

        expect(wrapper.find('.accordion-collapse').attributes('data-bs-parent')).toBe('#i-am-your-father');
    });
});
