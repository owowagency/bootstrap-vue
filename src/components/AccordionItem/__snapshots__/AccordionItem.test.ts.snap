// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`template > renders default 1`] = `
"<div class=\\"accordion-item\\">
  <h2 class=\\"accordion-header\\"><button class=\\"accordion-button collapsed\\" data-bs-toggle=\\"collapse\\" data-bs-target=\\"#dont-change-please\\">
      <!-- @slot Displays the header of the accordion -->
    </button></h2>
  <div id=\\"dont-change-please\\" class=\\"accordion-collapse collapse\\">
    <div class=\\"accordion-body\\">
      <!-- @slot Displays the body of the accordion -->
    </div>
  </div>
</div>"
`;

exports[`template > renders given custom classes 1`] = `
"<div class=\\"accordion-item custom-class\\">
  <h2 class=\\"accordion-header header-class\\"><button class=\\"accordion-button collapsed button-class\\" data-bs-toggle=\\"collapse\\" data-bs-target=\\"#owow-accordion-item\\">
      <!-- @slot Displays the header of the accordion -->
    </button></h2>
  <div id=\\"owow-accordion-item\\" class=\\"accordion-collapse collapse collapse-class\\">
    <div class=\\"accordion-body body-class\\">
      <!-- @slot Displays the body of the accordion -->
    </div>
  </div>
</div>"
`;

exports[`template > renders header 1`] = `
"<div class=\\"accordion-item\\">
  <h5 class=\\"accordion-header\\"><button class=\\"accordion-button collapsed\\" data-bs-toggle=\\"collapse\\" data-bs-target=\\"#owow-accordion-item\\">
      <!-- @slot Displays the header of the accordion -->find-me
    </button></h5>
  <div id=\\"owow-accordion-item\\" class=\\"accordion-collapse collapse\\">
    <div class=\\"accordion-body\\">
      <!-- @slot Displays the body of the accordion -->
    </div>
  </div>
</div>"
`;
